Downgraded xdlops to feature::mfma

zhanglx13 opened the pull request:
Currently, we have xdlops as the ONLY advanced feature when generating kernels. To prepare us with more than one features, such as the `wmma` instructions in Navi3x GPUs, this PR refactored the command line options of `rocmlir-gen` so that developers can control each feature during kernel generation.

- Added options for `rocmlir-gen`:
    - `--arch <arch>` is added as a required option if `rocmlir-gen` gets its inputs from `cl::opt`. This is not required if `rocmlir-gen` gets its inputs from user defined kernel or `--conv_config`. Our tests obtain the `arch` information from `rocm_agent_enumerator -name` invoked at lit configuration time.
    - For each feature, developers can use `<feature>=on` (force enable the feature), `<feature>=off` (force disable the feature), and `<feature>=infer` (only enable if the chip supports the feature) to control the feature list of the generated kernel. By default, `infer` is used.
- An argument is added to `Conv2dGenerator::isApplicable()` to allow us to skip the chip check. This is necessary when we want to generate xdlops kernels on non-xdlops GPUs.
- A cmake macro `ROCMLIR_GEN_FLAGS` is used to replace the old `ROCMLIR_DRIVER_XDLOPS_TEST_ENABLED`
    - The value of `ROCMLIR_GEN_FLAGS` is passed directly to `rocmlir-gen` in our E2E tests. For now, the only usage of this macro is to disable xdlops tests on gfx908 and gfx90a GPUs. Later, we can use this macro to toggle between similar features on the same GPU.
    - `ROCMLIR_DRIVER_XDLOPS_TEST_ENABLED` is dropped since xdlops tests are enabled automatically on GPUs that support xdlops and disabled on GPUs that do not. The decision is made based on chip information.
- In tests, feature list and arch attribute are fixed so that they match what has been set by `--arch <arch>` and `<feature>=on/off/infer`.

In the following PR, we can propagate the `arch` information from `rocmlir-gen` all the way to `mlir-rocm-runner` so that we can eliminate `hipGetDeviceProperties()` at runtime.

zhanglx13 committed:
Downgraded xdlops to feature::mfma

Instead of set -x2 to enable xdlops for rocmlir-gen, this PR makes the
advanced features default on each chip. To disable the advanced
feature, developer can set -no-<featureName> when invoking
rocmlir-gen.
Now the only supported feature is mfma.

TODO:
- Update e2e tests
- Update jenkinsfile

krzysz00 commented on CMakeLists.txt:
```diff
@@ -30,6 +30,7 @@ set(ROCMLIR_DRIVER_MISC_E2E_TEST_ENABLED 0 CACHE BOOL "Enable build miscellaneou
 set(ROCMLIR_DRIVER_PR_E2E_TEST_ENABLED 1 CACHE BOOL "Enable build PR-triggered E2E tests for Rock driver")
 set(ROCMLIR_DRIVER_RANDOM_DATA_SEED "none" CACHE STRING "Enable E2E tests using random data")
 set(ROCMLIR_DRIVER_XDLOPS_TEST_ENABLED 0 CACHE BOOL "Enable build E2E tests for XDLops")
+set(ROCMLIR_DRIVER_DISABLE_MFMA 1 CACHE BOOL "Disable build E2E tests for XDLops")
```
I'd set this to `0` so that the CI generally runs with mfma enabled

krzysz00 replied:
Especially given the comment on the README below

krzysz00 commented on mlir/test/CMakeLists.txt:
```diff
@@ -45,6 +46,14 @@ if(ROCMLIR_DRIVER_XDLOPS_TEST_ENABLED)
+# Passed to lit.site.cfg.py.in to set the -no-mfma flag for the mlir E2E tests.
+set(ROCMLIR_DISABLE_ADVANCED_FEATURE "")
+if(ROCMLIR_DRIVER_DISABLE_MFMA)
+  set(ROCMLIR_DISABLE_ADVANCED_FEATURE "-no-mfma")
```
Maybe two dashes for everyone's sanity?

krzysz00 commented on mlir/test/CMakeLists.txt:
```diff
@@ -45,6 +46,14 @@ if(ROCMLIR_DRIVER_XDLOPS_TEST_ENABLED)
 
+# Passed to lit.site.cfg.py.in to set the -no-mfma flag for the mlir E2E tests.
+set(ROCMLIR_DISABLE_ADVANCED_FEATURE "")
+if(ROCMLIR_DRIVER_DISABLE_MFMA)
```
Having stared at this, we might want to make this more general? Instead of having `_DISABLE_BLAH`, we could have `ROCMLIR_DRIVER_FORCE_BLAH` , which is by default unset. If you set it to ON, you get a `--blah` in your options (ex. `--mfma`), and setting it to OFF gives you `--no-blah`)

Then, in the drivers, you check to see if the boolean option has been set: if the uses passed neither `--blah` nor `--no-blah`, use the system default)

krzysz00 commented on mlir/test/e2e/lit.cfg.py:
```diff
@@ -33,6 +33,7 @@
 config.substitutions.append(("%mlir_src_root", config.mlir_src_root))
 config.substitutions.append(('%random_data', config.random_data))
 config.substitutions.append(('%xdlops', config.xdlops))
+config.substitutions.append(('%no_advanced_feature', config.no_advanced_feature))
```
And, if we don't go with the above generality comment, perhaps name this something like `%force_features` or `%disable_features`?

krzysz00 reviewed (commented):
I think there might be a more general framework here

zhanglx13 committed:
Replaced %xdlops with %feature

rocmlir-gen
- It uses the --feature option to determine the chip,
  which can be used to search for default GemmFeatures.
  The --feature to chip to AmdArchInfo is listed below:
  - mfma --> gfx908 --> cdnaInfo
  - pre-wmma --> gfx1030 --> rdnaInfo
  - wmma --> gfx1100 --> gfx11Info
  - vanilla --> gfx906 --> rdnaInfo (default)
  - none --> gfx900 --> gcnInfo
- The default feature is set to vanilla, which includes the dot
product (will be implemented in the future) and corresponds to
non-xdlops in our old terms.

From jenkins to rocmlir-gen
- `ROCMLIR_DRIVER_FEATURE=<feature>` is used to control
  - whether xdlops tests are supported
  - the --feature option passed to rocmlir-gen

krzysz00 reviewed (changes requested):
Having skimmed the description, this isn't what I wanted, and I'm sorry for the miscommunication.

I want there to be the following options:
- `--mfma` 
- `--no-mfma`
- `--dot`
- `--no-dot`
- `--atomic-add`
- `--no-atomic-add`

If neither the `--thing` or `--no-thing` options are set, then the default for the architecture you're running on is used for that feature

krzysz00 commented:
This may also involve going in, ripping the changes to `lit.*` out of https://github.com/ROCmSoftwarePlatform/rocMLIR/pull/762/files, and then actually setting the `--arch` argument to `rocmlir-gen` ... which is probably the right place to put that value, especially if we can then pick it up in `mlir-rocm-runner` by setting the value of `--arch` on a module as an attribute or something ... or just passing it again

zhanglx13 commented:
1. Re `--thing` and `--no-thing` options. I found it hard to set the `GemmFeatures` with these options. Do you expect only one of these options or none is passed to `rocmlir-gen`? 
    - If so, it is confusing to have both `--thing` and `--no-thing` options, since `--thing` seems to be adding a feature from an empty set while `--no-thing` seems to be removing a feature from a complete set.
    - If you suggest that `rocmlir-gen` obtains the default features according to the chip first, and then adds or removes a feature from the default features according to these `--thing` or `--no-thing` options, it is non-trivial to disable xdlops on gfx908||gfx90a GPUs. Because in this case, you need to remove multiple features from the default feature list.
    - If you allow multiple options in the form of `--things` or `--nothings`, it is like users of `rocmlir-gen` are constructing the `GemmFeatures` manually, which is error-prone and unclear of what pipeline we want to generate. 
2. Re `--arch`. The old way for `rocmlir-gen` to figure out the chip is according to whether `-x2` is set. If so, choose `gfx908` and if not, choose `gfx900`. In the comment of `populateDefaults()`, it says that `rocmlir-gen` does not care about the chip/arch since `rocmlir-lib` and `mlir-rocm-runner` both have their own ways to obtain the chip information. So are you suggesting that `rocmlir-gen` also needs to obtain the correct chip information from the user input?

In the good old days, where `mfma` is the only feature, it is simple and clear to use a single option `-x2` to indicate if we want to enable xdlops. However, when we have multiple features, we will have multiple generation pipelines, such as `mfma`, `wmma`, and `dot` (or `vanilla` after `gfx900` is officially dropped). Essentially, the `--feature` option to `rocmlir-gen` is to choose the pipeline, which does not match exactly to the features in `GemmFeatures`. (maybe we should rename it to something not confused with `GemmFeatures`).

zhanglx13 committed:
Replaced arch = gfx900 with gfx906 for non-xdlops

krzysz00 commented:
1. ... you have a point. It might be better to use https://llvm.org/docs/CommandLine.html#selecting-an-alternative-from-a-set-of-possibilities to have, for instance, `--infer-mfma`, `--enable-mfma`, and `--disable-mfma` - heck, you could probably use the same enum for each feature, since the cases are the same, just the names are different.
  - The point of these options is to let us, the developers, explicitly toggle the defaults. The starting point in the results of the architecture DB lookup, and then we explicitly set or clear (we have `bitEnumSet(bits, bit, bool value)` for this purpose if the user calls for it
3. I claim that, to fulfill a long-standing low-priprity ticket, we should stop doing chip detection in `mlir-rocm-runner`. Instead, we should grab `--arch` in the test system (roughly how I went for `--targets` in the PR that wasn't working out). Since we need that information in `rocmlir-gen`, it should be fed in there, and then propagated down the pipeline until it reaches `mlir-rocm-runner` (which can probably keep the detection code just in case, but still).
4. However, as a short term matter, we could
  - Default "arch" to "gfx900" or something else that won't cause problems to matter what (since we don't rely on wavesize anywhere), allowing the feature inference to 'work'
  - Make what used to be `-x2` into `--enable-mfma --enable-dot --enable-atomic-add` to match the existing architectures
  - (This is definitely in the realm of a quick hack, but it'll let us make `--arch` a separate PR)
5. What I don't want to see is fake keys into the architecture database that we start using for CI purposes. If we're going to be selecting features, we need to go with a real `--arch` that matches the system you're actually running on.

zhanglx13 committed:
Added --arch %targetChip to rocmlir-gen when it's using cl::opt to
construct the kernel

%targetChip is obtained from rocm_agent_enumerator during lit setup.
Changes are copied from https://github.com/ROCmSoftwarePlatform/rocMLIR/pull/762

zhanglx13 committed:
Merge branch 'miopen-dialect' into downgrade-xdlops-rocmlir-gen

zhanglx13 committed:
Added options for each feature

so that developers can manually add or remove each feature into/from
the default feature list.
If nothing is specified or infer is chose, the default value from the
underlying chip is used.

zhanglx13 committed:
Added cmake macros to force each feature

zhanglx13 committed:
Merge branch 'miopen-dialect' into downgrade-xdlops-rocmlir-gen

zhanglx13 committed:
Fixed tests

- -feature=mfma is replaced with -mfma=on
- rocmlir-gen flag matches feature attributes of kernels

zhanglx13 committed:
Skip chip check when invoking rocmlir-gen

This fixed failed tests on gfx1030

zhanglx13 committed:
Updated Jenkinsfile

krzysz00 commented on CMakeLists.txt:
```diff
@@ -30,6 +30,10 @@ set(ROCMLIR_DRIVER_MISC_E2E_TEST_ENABLED 0 CACHE BOOL "Enable build miscellaneou
 set(ROCMLIR_DRIVER_PR_E2E_TEST_ENABLED 1 CACHE BOOL "Enable build PR-triggered E2E tests for Rock driver")
 set(ROCMLIR_DRIVER_RANDOM_DATA_SEED "none" CACHE STRING "Enable E2E tests using random data")
 set(ROCMLIR_DRIVER_XDLOPS_TEST_ENABLED 0 CACHE BOOL "Enable build E2E tests for XDLops")
+set(ROCMLIR_DRIVER_FEATURE "mfma" CACHE BOOL "Choose the feature")
```
These should be defaulted to empty string (or "none" or what have you) and the booleans below should default to 0. The build should work by default on a non-xdlops scenario

krzysz00 commented on README.md:
```diff
@@ -17,9 +17,13 @@ cmake -G Ninja .. -DCMAKE_BUILD_TYPE=RelWithDebInfo
-If you will be targetting a MI-100, MI-200, or other system that supports
-mfma instructions, add the flag `-DROCMLIR_DRIVER_XDLOPS_TEST_ENABLED=1 `
-to the `cmake` invocation above.
+By default, tests requiring xdlops are enabled, which requires
```
xdlops tests are to be disabled by default

krzysz00 commented on mlir/test/CMakeLists.txt:
```diff
@@ -45,6 +48,29 @@ if(ROCMLIR_DRIVER_XDLOPS_TEST_ENABLED)
+set(ROCMLIR_FEATURE "")
+# mfma feature
+if(ROCMLIR_DRIVER_FORCE_MFMA)
+  set(ROCMLIR_FEATURE "-mfma=on")
```
Can we make this uniform with the other ones, where they take `${ROCMLIR_FEATURE}`?

Come to think it, can we make this be `ROCMLIR_GEN_FLAGS` and have it be a user-controllable feature so we don't have all the booleans?

Or we can keep the booleans as well. I'd say `ROCMLIR_GEN_FLAGS` as a CMake testing setup is perhaps a good call

krzysz00 commented on mlir/test/e2e/lit.cfg.py:
```diff
@@ -33,6 +33,8 @@
 config.substitutions.append(("%mlir_src_root", config.mlir_src_root))
 config.substitutions.append(('%random_data', config.random_data))
 config.substitutions.append(('%xdlops', config.xdlops))
+config.substitutions.append(('%feature', config.feature))
```
Let's call this `%features` or `%flags`?

krzysz00 commented on mlir/test/e2e/lit.site.cfg.py.in:
```diff
@@ -36,13 +38,16 @@ except KeyError:
 # If rocm_agent_enumerator shows no viable GPUs, skip tests that need one,
 # because the default target will lead to compilation failures.
 config.no_AMD_GPU = False
+config.targetChip = ""
```
Could you please incorporate the `-name` and no `-name` structure from https://github.com/ROCmSoftwarePlatform/rocMLIR/pull/762 ?

krzysz00 commented on mlir/tools/rocmlir-gen/rocmlir-gen.cpp:
```diff
@@ -2183,20 +2210,30 @@ int main(int argc, char **argv) {
-      std::string triple = splitter.getTriple().str();
-      std::string chip = splitter.getChip().str();
-      std::string chipFeatures = splitter.getFeatures().str();
+      std::string chip = arch.getValue();
```
This should be using the arch name splitter

krzysz00 replied:
Since the value of `arch` will be `chip:features`

krzysz00 commented on mlir/tools/rocmlir-gen/rocmlir-gen.cpp:
```diff
@@ -223,6 +224,45 @@ static cl::opt<bool>
              cl::init(false));
 
+// disable feature mfma
+static cl::opt<std::string> featureInfo(
```
I'd like to not call this "feature".

If we want this option at all, which I'm not sure we do, I want this to be something like `codepath`

krzysz00 reviewed (commented):
I've got mainly naming comments

Overall, one option might be to drop the booleans and have the interface be `ROCMLIR_GEN_FLAGS`, perhaps with a convenience boolean (XDLOPS_TEST_ENABLED is still good for it)

Also, the "feature" option in rocmlir-gen isn't being checked anywhere, we might as well drop it.

zhanglx13 committed:
Fixed label mismatch on gfx90a GPUs

zhanglx13 committed:
Simplified the macros for controlling features

zhanglx13 committed:
%feature --> %features

krzysz00 commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -256,7 +256,7 @@ pipeline {
                                 buildProject('check-rocmlir',
                                              """-DROCMLIR_DRIVER_PR_E2E_TEST_ENABLED=0
-                                             -DROCMLIR_DRIVER_XDLOPS_TEST_ENABLED=${params.canXdlops ? '1' : '0'}
+                                             -DROCMLIR_GEN_FEATURES=${params.canXdlops ? '-mfma=on' : '-mfma=off'}
```
This should be `--mfma=on --atomic_add=on` and similarly for the off

krzysz00 reviewed (commented):
One minor bit of weirdness wrt needing to also set the atomic_add flags on aside, works well

krzysz00 commented:
MY other nitpick is that it should be `%arch` and not `%targetChip`

zhanglx13 commented:
Yeah. I was missing the fact that `rocm_agent_enumerator -name` can also include features.

krzysz00 commented:
No problem.
Having the features in there is the point.
The idea is that we pass in `%arch` at the beginning (someone'll have a followup PR to bounce it along) and then we don't have to do any `hipGetDeviceProperties()` calls

zhanglx13 committed:
%targetChip --> %arch

It includes chip:features

krzysz00 commented on mlir/test/rocmlir-driver/auto_e2e/xdlops/lit.local.cfg:
```diff
@@ -1,2 +1,2 @@
-if not config.enable_rock_driver_e2e_test or config.no_AMD_GPU or not config.xdlops:
+if not config.enable_rock_driver_e2e_test or config.no_AMD_GPU or "-mfma=on" not in config.features:
```
For this, I'd suggest something different.
Specifically, what I'd go for is to, over in `lit.site.cfg.py.in`, look through `%arch` to check if it contains `gfx908`, `gfx90a`, or so on, and, if it does, set `config.arch_has_mfma = True`

krzysz00 commented on mlir/utils/jenkins/Jenkinsfile.downstream:
```diff
@@ -94,7 +94,9 @@ pipeline {
                                   -DROCMLIR_DRIVER_PR_E2E_TEST_ENABLED=1
-                                  -DROCMLIR_DRIVER_XDLOPS_TEST_ENABLED=0
+                                  -DROCMLIR_DRIVER_FORCE_MFMA=OFF
+                                  -DROCMLIR_DRIVER_FORCE_DOT=OFF
```
Per before, these flags don't seem to exist anymore

krzysz00 commented on mlir/utils/jenkins/Jenkinsfile:
```diff
@@ -256,7 +256,7 @@ pipeline {
                                 buildProject('check-rocmlir',
                                              """-DROCMLIR_DRIVER_PR_E2E_TEST_ENABLED=0
-                                             -DROCMLIR_DRIVER_XDLOPS_TEST_ENABLED=${params.canXdlops ? '1' : '0'}
+                                             -DROCMLIR_GEN_FEATURES=${params.canXdlops ? '-mfma=on -atomic_add=on' : '-mfma=off --atomic_add=off'}
```
Given that we're passing `%arch`, we don't _need_ to set this.
I can see leaving it in, just so that when the multi-feature PR lands, we have the hook to customize it, but we don't actually have to set `-mfma=on`

krzysz00 reviewed (commented):
Thanks!

I think the tests for `xdlops` directories should rely on a `config.arch_has_mfma` derived setting instead of querying on the features flag, since that means that we won't be building those tests in our local dev environments where everything is inferred.

(I also think we can avoid explicitly setting `ROCMLIR_GEN_FEATURES` at all, but it probably can't hurt to do it)

zhanglx13 commented:
You're right. Actually having `%features` and having `nothing` are the same, since they both infer the feature according to the chip. Then Jenkins does not need to do that and can leave the job to `rocmlir-gen`.

krzysz00 commented:
I'd leave %features in anyway - you'll want it for your multi-codepath testing PR

krzysz00 commented:
Could name it `%rocmlir_gen_flags` by analogy with `%cxxflags` and the like

zhanglx13 commented:
Why it is needed for the multi-codepath PR? Jenkins will pick the correct label according to the codepath and then `rocmlir-gen` can infer the default features according to the chip.

krzysz00 commented:
We might want to explicitly test `--dot=off`

zhanglx13 committed:
Guard xdlops test dir from config.arch

Also %features --> %rocmlir_gen_flags

krzysz00 reviewed (approved):
This all looks pretty reasonable to me, approved

zhanglx13 committed:
ROCMLIR_GEN_FEATURES --> ROCMLIR_GEN_FLAGS

And cleanups

krzysz00 reviewed (changes requested):
Wait, no, I just noticed: the performance scripts need to pass %arch now
And to stop passing `-x2`

zhanglx13 committed:
Fixed rocmlir-gen commands in MIOpenDriver.py

krzysz00 reviewed (commented):
I'll note that `parameterSweeps.py` has a fork of the class you modified, you'll need to get that too

zhanglx13 committed:
Fixed rocmlir-gen commands in parameterSweeps.py

zhanglx13 committed:
Added -mfma=on/off to rocmlir-gen in MIOpenDriver.py and parameterSweeps.py

krzysz00 reviewed (approved):
Having caught that last-minute breakage potential .... approved

zhanglx13 committed:
Merge branch 'miopen-dialect' into downgrade-xdlops-rocmlir-gen

zhanglx13 committed:
Merge branch 'miopen-dialect' into downgrade-xdlops-rocmlir-gen

zhanglx13 committed:
Fixed clang-format issues

zhanglx13 committed:
Merge branch 'miopen-dialect' into downgrade-xdlops-rocmlir-gen
